pkg/controller/cronjob/utils_test.go

```diff
@@ -268,114 +268,126 @@ func TestGetRecentUnmetScheduleTimes(t *testing.T) {
+			// Status shows a start at the expected time.
+			lastScheduleTime: &metav1.Time{Time: T1},
+			deadline:         nil,
+			expectErr:        false,
+			expectResult:     []time.Time{},
+		},
+		{
+			caseName: "Case 4: known LastScheduleTime, a start needed.",
+			// Current time is after T1 and after T2
+			nowTime: T2.Add(5 * time.Minute),
+			// Creation time is before T1.
+			creationTimestamp: metav1.Time{Time: T1.Add(-10 * time.Minute)},
+			// Status shows a start at the expected time.
+			lastScheduleTime: &metav1.Time{Time: T1},
+			deadline:         nil,
+			expectErr:        false,
+			expectResult:     []time.Time{T2},
+		},
+		{
+			caseName: "Case 5: known LastScheduleTime, two starts needed.",
+			// Current time is after T1 and after T2
+			nowTime:           T2.Add(5 * time.Minute),
+			creationTimestamp: metav1.Time{Time: T1.Add(-2 * time.Hour)},
+			lastScheduleTime:  &metav1.Time{Time: T1.Add(-1 * time.Hour)},
+			deadline:          nil,
+			expectErr:         false,
+			expectResult:      []time.Time{T1, T2},
+		},
+		{
+			caseName:          "Case 6: now is way way ahead of last start time, and there is no deadline.",
+			nowTime:           T2.Add(10 * 24 * time.Hour),
+			creationTimestamp: metav1.Time{Time: T1.Add(-2 * time.Hour)},
+			lastScheduleTime:  &metav1.Time{Time: T1.Add(-1 * time.Hour)},
+			deadline:          nil,
+			expectErr:         true,
+			expectResult:      []time.Time{},
+		},
+		{
+			caseName:          "Case 7: now is way way ahead of last start time, but there is a short deadline",
+			nowTime:           T2.Add(10 * 24 * time.Hour),
+			creationTimestamp: metav1.Time{Time: T1.Add(-2 * time.Hour)},
+			lastScheduleTime:  &metav1.Time{Time: T1.Add(-1 * time.Hour)},
+			// Deadline is short
+			deadline:     &deadline,
+			expectErr:    false,
+			expectResult: []time.Time{T1.Add(240 * time.Hour), T1.Add(241 * time.Hour)},
+		},
 	}
-	{
-		// Case 4: known LastScheduleTime, a start needed
-		// Creation time is before T1.
-		sj.ObjectMeta.CreationTimestamp = metav1.Time{Time: T1.Add(-10 * time.Minute)}
-		// Status shows a start at the expected time.
-		sj.Status.LastScheduleTime = &metav1.Time{Time: T1}
-		// Current time is after T1 and after T2
-		now := T2.Add(5 * time.Minute)
-		times, err := getRecentUnmetScheduleTimes(sj, now)
-		if err != nil {
-			t.Errorf("unexpected error: %v", err)
-		}
-		if len(times) != 1 {
-			t.Errorf("expected 1 start times, got: %v", times)
-		} else if !times[0].Equal(T2) {
-			t.Errorf("expected: %v, got: %v", T1, times[0])
+
+	for _, testCase := range testCases {
+		sj.ObjectMeta.CreationTimestamp = testCase.creationTimestamp
+		sj.Status.LastScheduleTime = testCase.lastScheduleTime
+		sj.Spec.StartingDeadlineSeconds = testCase.deadline
+
+		times, err := getRecentUnmetScheduleTimes(sj, testCase.nowTime)
+		if err != nil && !testCase.expectErr {
+			t.Errorf("unexpect getRecentUnmetScheduleTimes err: %v, caseName: %v", err, testCase.caseName)
```

Comment:
When using `t.Run` you won't have to deal with `caseName` in the logs.

Reply:
Already fixed.

Reply:
Already fixed.